File: Algo/Easy/0020_Valid_Parentheses/Solution.py
from typing import *

# 2022-03-12
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) == 1:
            return False
        
        stack = []
        leftBrackets = {'{': '}', '[': ']', '(': ')'}
        
        for bracket in s:
            if bracket in leftBrackets:
                stack.append(bracket)
            else:
                if len(stack) == 0:
                    return False
                
                previous = stack.pop()
                if bracket != leftBrackets[previous]:
                    return False
        
        return len(stack) == 0

File: Algo/Easy/0020_Valid_Parentheses/Solution.cpp
// 2022-08-31 (stack)
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> brackets = {
            {'(', ')'},
            {'[', ']'},
            {'{', '}'},
        };
        
        deque<char> stack;
        for (int i=0; i<s.length(); i++) {
            if (brackets.count(s[i]) != 0) {
                stack.push_back(s[i]);
            } else {
                if (stack.size() == 0 || brackets[stack.back()] != s[i])
                    return false;
                
                stack.pop_back();
            }
        }
        
        return stack.size() == 0;
    }
};